docs/migrating.rst

```diff
@@ -0,0 +1,282 @@
+implementation, and sometimes even to remove existing functionality
+altogether.
+
+One of the main objectives of this rewrite was to enable `ndb` for use in any
+Python environment, not just Google App Engine. As a result, many of the `ndb`
+APIs that relied on GAE environment and runtime variables, resources, and
+legacy APIs have been dropped.
+
+Aside from this, there are many differences between the Datastore APIs
+provided by GAE and those provided by the newer Google Cloud Platform. These
+differences have required some code and API changes as well.
+
+Finally, in many cases, new features of Python 3 have eliminated the need for
+some code, particularly from the old `utils` module.
+
+If you are migrating code, these changes can generate some confusion. This
+document will cover the most common migration issues.
+
+Setting up a connection
+=======================
+
+The most important difference from the previous `ndb` version, is that the new
+`ndb` requires the use of a client to set up a runtime context for a project.
+This is necessary because `ndb` can now be used in any Python environment, so
+we can no longer assume it's running in the context of a GAE request.
+
+The `ndb` client uses ``google.auth`` for authentication, which is how APIs in
+Google Cloud Platform work. The client can take a `credentials` parameter or
+get the credentials using the `GOOGLE_APPLCATION_CREDENTIALS` environment
+variable, which is the recommended option.
+
+After instantiating a client, it's necessary to establish a runtime context,
+using the ``Client.context`` method. All interactions with the database must
+be within the context obtained from this call::
+
+    from google.cloud import ndb
+
+    client = ndb.Client()
+
+    with client.context() as context:
+        do_something_with_ndb()
+
+The context is not thread safe, so for threaded applications, you need to
+generate one context per thread. This is particularly important for web
+applications, where the best practice would be to generate a context per
+request.
+
+The following code shows how to use the context in a threaded application::
+
+    import threading
+    from google.cloud import datastore
+    from google.cloud import ndb
+
+    client = ndb.Client()
+
+    class Test(ndb.Model):
+        name = ndb.StringProperty()
+
+    def insert(input_name):    
+        with client.context():
+            t = Test(name=input_name)        
+            t.put()        
+
+    thread1 = threading.Thread(target=insert, args=['John'])
+    thread2 = threading.Thread(target=insert, args=['Bob'])
+
+    thread1.start()
+    thread2.start()
+
+Note that the examples above are assuming the google credentials are set in
+the environment.
+
+Keys
+====
+
+There are some methods from the ``key`` module that are not implemented in
+this version of `ndb`:
+
+    - Key.from_old_key.
+    - Key.to_old_key.
+
+These methods were used to pass keys to and from the db Datastore API, which is
```

Comment:
Should make obvious that ``db`` is a name of an old product. In this case, the predecessor to ``ndb``.